Add chunking support to vectorize.table()

asr2003 opened the pull request:
Closes #142 
/claim #142

asr2003 commented:
@ChuckHend Have a look at your free time and kindly approve workflows, if any adjustments needed i am happy to incorporate them

asr2003 commented:
Waiting for a quick review on this too - @ChuckHend

asr2003 commented:
@ChuckHend Updated the changes with intended approach of chunking `vectorie.table` by retreiving rows and lives into new table,

Have a review on it and If any changes or modifications required I am ready to incorporate them

asr2003 commented:
@ChuckHend Have a review on it in your free time

asr2003 commented:
Sure! I will update the tests

asr2003 commented:
Just a minor query just does we need automatic chunking once table is
called and it sync with chunk_table?

On Mon, 4 Nov, 2024, 8:43 pm Adam Hendel, ***@***.***> wrote:

> Can you add a test or two that shows how the functionality will work and
> assert that it functions as expected?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tembo-io/pg_vectorize/pull/162#issuecomment-2454977061>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BGXZB6GUWQYZTX2ZY75W7ATZ656JHAVCNFSM6AAAAABQEOKG32VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINJUHE3TOMBWGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ChuckHend commented:
Yes, let's leave the integration with `vectorize.table()` out to make this simpler / easier. Just keep chunk_text() a sql function that we can call directly, and add a couple tests that validate its functionality and shows how to use it.

asr2003 commented:
Okay!

asr2003 commented:
Just need test update and I am confused of I am going in right way of testing that function @ChuckHend

asr2003 commented:
@ChuckHend Any reviews on it

ChuckHend commented:
Ill have this reviewed by end of week

ChuckHend commented on extension/src/util.rs:
```diff
@@ -2,7 +2,8 @@ use anyhow::Result;
 use pgrx::*;
 use sqlx::postgres::{PgConnectOptions, PgPoolOptions};
-use sqlx::{Pool, Postgres};
+use sqlx::{Pool, Postgres, Row};
+use sqlx::{Pool, Postgres, Row};
```
there's duplicate imports here

ChuckHend commented on extension/src/api.rs:
```diff
@@ -167,3 +167,40 @@ fn env_interpolate_guc(guc_name: &str) -> Result<String> {
+        }
+
+        // Fallback if no suitable separator is found, chunk by size
+        if !chunk_found {
```
`chunk_found` is not defined anywhere, so this fails

ChuckHend commented on extension/src/api.rs:
```diff
@@ -167,3 +167,40 @@ fn env_interpolate_guc(guc_name: &str) -> Result<String> {
+            if let Some(pos) = text[start..end].rfind(sep) {
+                if pos > 0 {
+                    end = start + pos;
+                    chunks.push(chunk.clone());
```
`chunk` is not defined

asr2003 commented:
I haven't pushed my latest changes here which are in my local :)

asr2003 commented:
@ChuckHend We are mostly there and need fixing of tests
```
thread 'test_chunk_text' panicked at tests/integration_tests.rs:68:5:
assertion `left == right` failed
  left: ["This is a", " test for", " chunking.", "king."]
 right: ["This is a ", "is a test ", "test for c", "for chunki", "chunking."]
```

asr2003 commented:
@ChuckHend 👋
I have updated the `chunk_text` function to use the `text_splitter` crate instead of the previous manual approach. This change makes the chunking more accurate especially for larger documents and ensures we preserve word boundaries while still enforcing a character limit. The previous logic sometimes caused inconsistencies with expected vs actual outputs. Using `text_splitter` brings more stability and aligns with the best practices seen in LangChain’s recursive text splitter. I have also updated docs of it's usage as standalone  sql function and added couple of tests that validate its functionality and tests passes locally. 

With this I think it's ready to merge

ChuckHend commented on extension/src/util.rs:
```diff
@@ -212,3 +212,17 @@ pub async fn ready(conn: &Pool<Postgres>) -> bool {
 }
+
+/// Fetch rows from a given table and schema
+pub async fn fetch_table_rows(
```
is this function being used anywhere?

asr2003 replied:
Nope! It used in previous approach, will remove it
